improve build.zig + use v0.14.0 idioms

Torbatti opened the pull request:
- added build steps and build options
this will make finding what steps are used and what options are available easier

- you only need zig build and zig build run 
specifiy examples to build/run by using `-Dexample=[nameOfExample]` option

- added static step to build tardy as a standalone lib `zig build static`
- e2e test been ported `zig build test_e2e -Dasync=auto`
-  unit tests been ported `zig build test_unit`
- added formatting checker test `zig build test_fmt`
- `zig build test` will now run unit tests and fmting tests

- decoupled build example module and build example exe functions
- replaced root_module with the new createMod and addImport introduced in v0.14.0
- note: github actions wont compile any examples for os piplines  / can be added via run step `zig build -Dexample=all`

mookums commented on build.zig:
```diff
@@ -16,111 +17,371 @@ comptime {
+        inline for (std.meta.fields(Example)) |f| {
+
+            // skip .none and .all for building step
+            if (f.value == 1 or f.value == 0) {
```
I'd probably use the enum name here instead of the values.

mookums commented on build.zig:
```diff
@@ -16,111 +17,371 @@ comptime {
+    stat,
+    stream,
+
+    fn toString(ex: Example) []const u8 {
```
Instead of this, why not just use `@tagName` on the enum? You would get the same strings and avoid a bespoke conversion function.

Torbatti replied:
there are some advantages for using `toString` helper function instead of `@tagName` 
 1. use nested paths and get a simpler enum field name
 2. put similar examples inside a single directory
```
using toString:
examples/echo/v1/ -> .echo_v1
examples/echo/v2/ -> .echo_v2

using @tagName
examples/echo_v1/ -> .echo_v1
examples/echo_v2/ -> .echo_v2
```
for disadvantages :
 may forget to add example to toString helper function

mookums commented on build.zig:
```diff
@@ -16,111 +17,371 @@ comptime {
+
+    // create a private example module
+    const example_mod = b.createModule(.{
+        .root_source_file = b.path(b.fmt("examples/{s}/main.zig", .{options.example.toString()})),
```
`@tagName(options.example)`

mookums commented on build.zig:
```diff
@@ -16,111 +17,371 @@ comptime {
+    steps.install.dependOn(&install_artifact.step);
+
+    // Should not run all examples at the same time
+    if (options.allExamples) {
```
Should log that you can't run all examples at the same time.

mookums commented on build.zig:
```diff
@@ -16,111 +17,371 @@ comptime {
+    });
+
+    const example_exe = b.addExecutable(.{
+        .name = options.example.toString(),
```
`@tagName(options.example)`

mookums commented on build.zig:
```diff
@@ -16,111 +17,371 @@ comptime {
-const AsyncKind = @import("src/aio/lib.zig").AsyncKind;
+fn build_static_lib(
+    b: *std.Build,
+    steps: struct {
```
No reason to have this be a struct.

mookums commented on build.zig:
```diff
@@ -16,111 +17,371 @@ comptime {
+    options: struct {
+        tardy_mod: *std.Build.Module,
+        example: Example,
+        allExamples: bool,
```
allExamples -> all_examples

mookums commented on README.md:
````diff
@@ -31,10 +31,31 @@ const tardy = b.dependency("tardy", .{
 ```
 
-## Example
+## Building and running example folder
````
Building and Running Examples

mookums commented:
Most of it is pretty good, left a few comments of things that might be worth taking a look at.

Torbatti commented:
- convert captured field value to field enum for ease of read
- changed campleCase allExamples  to snake_case all_examples
- add info log: let user know` zig build run -Dexample=all` will only build examples and will not runt them
- added usage comments inside main build function
